### ISSUE TITLE
regression in 3.0.3: strcasecmp_l is not working when using glibc 2.25 or older

### ISSUE THREAD

**Issue by ngg** [openssl/openssl#18244]
Before glibc 2.26 (fix was in commit https://github.com/bminor/glibc/commit/7b037c095e31c2396d0a9b0e6356bc566ee4812f), the `strcasecmp_l` function was defined in two headers: `<string.h>` and `<strings.h>`.
However, the former was only defined if `_GNU_SOURCE` is used and the latter used the proper POSIX check.
If you include both headers in this order, the symbol will not get defined if `_POSIX_C_SOURCE >= 200809L` but `_GNU_SOURCE` is not defined.

We hit exactly this case, when compiling OpenSSL on Centos 7 (x64):
```
crypto/ctype.c:355:12: warning: implicit declaration of function 'strncasecmp_l' is invalid in C99 [-Wimplicit-function-declaration]
    return strncasecmp_l(s1, s2, n, (locale_t)ossl_c_locale());
```

To fix this issue, I'd propose either of the following solutions:
- Define `_GNU_SOURCE` in `ctype.c`, there are some files in libcrypto that already do this
- Replace the order of the includes to include `<strings.h>` before `<string.h>`. This is uglier, but also works.

**Comment by rsbeckerca** [comment]
strings.h cannot be included before string.h on every platform, so the second suggestion would have negative consequences.

**Comment by t8m** [comment]
Could you please try the fixes in #18241?

**Comment by t8m** [comment]
> strings.h cannot be included before string.h on every platform, so the second suggestion would have negative consequences.

Really? I see we are including "e_os.h" which implicitly includes strings.h before string.h in multiple files.

**Comment by rsbeckerca** [comment]
I will try the latest set right now.

**Comment by rsbeckerca** [comment]
@t8m  Do you want me to remove #18233 at this stage?

**Comment by rsbeckerca** [comment]
> > strings.h cannot be included before string.h on every platform, so the second suggestion would have negative consequences.
> 
> Really? I see we are including "e_os.h" which implicitly includes strings.h before string.h in multiple files.

It looks like my info was out of date. This was a problem but not at this point. I think it is only an issue if `_XOPEN_SOURCE` is not supplied or has a very old value. I can dig deeper if you want but it is working in the OpenSSL build.

**Comment by t8m** [comment]
> I can dig deeper if you want but it is working in the OpenSSL build.

Not necessary to dig deeper.

**Comment by t8m** [comment]
> @t8m Do you want me to remove #18233 at this stage?

If you want, you can close it now.

**Comment by ngg** [comment]
> Could you please try the fixes in #18241?

I can confirm that the current changes of #18241 applied to the 3.0.3 version works in my case.
